### PR TITLE
driver: Add support for native fuzzing without launcher

### PR DESCRIPTION
**Pull request by fmeum**
Fuzzing native libraries with ASan and UBSan using the Java driver is
made possible by the generated launcher script, which this commit
uses to preload the Jazzer hooks and native sanitizer runtime libraries
into a JVM running Jazzer started as a subprocess.

Since the sanitizer instrumentation applied at compile time and the
runtime libraries must be kept in sync, the preload approach allows
users to fuzz native libraries without building the Jazzer launcher from
source for the first time. Jazzer will automatically pick the libraries
corresponding to `CC` (if set) or `clang` in `PATH`.

While the existing native_fuzzer_hooks.c can be reused almost unchanged
for this purpose on Linux, macOS requires mimicking the interposing
logic used by ASan to hook libc funtions at runtime. This commit
extracts these platform-specific hooking techniques into preprocessor
macros.

Further complications arise because macOS codesigning prevents library
insertion into `/bin/sh` via `DYLD_INSERT_LIBRARIES` and generally
requires all libraries to be preloaded into the `java` process to be
codesigned - users may have to click through Gatekeeper warnings to
allow this. This requires removing the signatures in CI, where we can't
simulate these clicks.

Tests fail on macOS 11 for unclear reasons, so we raise the minimum
version to 12, which hase been out for almost a year by now. They also
fail with JDK 8, which can't be reproduced locally - these tests are
skipped in CI.

### CODE REVIEW

**Review by bertschneider — COMMENTED** [review]
I really like this change. Preloading the sanitizers and so removing the specific jazzer variants makes the setup such simpler.

**Review by bertschneider — APPROVED** [review]

File: driver/src/main/native/com/code_intelligence/jazzer/jazzer_preload.c [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+                                 result);                              \
+    return result;                                                     \
+  }                                                                    \
+                                                                       \
+  INTERPOSE(interposed_##name, name)
+#else
+#error "jazzer_preload is not supported on this OS"
+#endif
+
+DEFINE_LIBC_INTERCEPTOR(bcmp, int, (const void *s1, const void *s2, size_t n),
```
Comment (kyakdan) [inline driver/src/main/native/com/code_intelligence/jazzer/jazzer_preload.c:130]: Could we get this to work for Windows using [detours](https://github.com/microsoft/Detours)?

File: driver/src/main/java/com/code_intelligence/jazzer/Jazzer.java [hunk 2]
```diff
@@ -116,10 +208,112 @@ private static Path javaBinary() {
...
+            // Presumably, this issue is caused by codesigning and the exec helper missing the
+            // entitlements required for library insertion.
+            "-Djdk.attach.allowAttachSelf=true", Jazzer.class.getName()));
+  }
+
+  /**
+   * Append the given elements to the value of the environment variable {@code name} that contains a
+   * list of paths separated by the system path list separator.
+   */
+  private static String appendWithPathListSeparator(String name, String... options) {
```
[the remark this answers is not in the document]
Comment (fmeum) [reply driver/src/main/java/com/code_intelligence/jazzer/Jazzer.java:229]: Good point. I added a precondition check that simplifies the logic even further down to just two cases. PTAL.

File: driver/src/main/java/com/code_intelligence/jazzer/Jazzer.java [hunk 3]
```diff
@@ -116,10 +208,112 @@ private static Path javaBinary() {
...
+      System.err.printf(
+          "ERROR: '%s' returned '%s', but it doesn't exist%n", String.join(" ", command), library);
+      exit(1);
+    }
+    return library;
+  }
+
+  private static String asanLibName() {
+    if (isLinux()) {
+      return "libclang_rt.asan-x86_64.so";
```
[the remark this answers is not in the document]
Comment (kyakdan) [reply driver/src/main/java/com/code_intelligence/jazzer/Jazzer.java]: I noticed this change with LLVM 15. The LLVM 14.0.0 release on GitHub also contains the libs in a new location (https://github.com/llvm/llvm-project/releases/tag/llvmorg-14.0.0)

File: driver/src/main/java/com/code_intelligence/jazzer/Jazzer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kyakdan) [reply driver/src/main/java/com/code_intelligence/jazzer/Jazzer.java:229]: Looks better!

**Review by kyakdan — APPROVED** [review]

**Pull request merged**